### PR TITLE
Update to latest DLR and upgrade to VS2017 and new csproj format

### PR DESCRIPTION
**Pull request by gpetrou**
Please ignore for now. Most csproj files are not yet converted.
Edit: Almost there

### CODE REVIEW

**Comment by gpetrou** [comment]
Hmm, how exactly should we change the make.cmd file to take into account the target framework?
Should we pass another parameter?

**Comment by slide** [comment]
Right now we build for all platforms in one go, I'd like to maintain that.

**Comment by gpetrou** [comment]
So, should I just do:
net40/IronPythonTest.exe --labels=All --result:all-result.xml
net45/IronPythonTest.exe --labels=All --result:all-result.xml
etc... in the make file?

**Comment by slide** [comment]
Yes, I agree. Don't worry about net40

**Comment by gpetrou** [comment]
So, should I drop net40 with my PR?
Maybe better to do that on a separate issue?

**Comment by slide** [comment]
Yes, separate PR. I'm crafting one right now.

**Comment by gpetrou** [comment]
OK thanks.
I will remove net40 from <TargetFrameworks>net40;net45</TargetFrameworks> in my PR.
Otherwise, I guess it will still build it.

**Comment by gpetrou** [comment]
I think some of the latest changes have affected my PR.
The packages.config should no longer be needed with the new format.

File: Src/IronPython.SQLite/IronPython.SQLite.csproj [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
+﻿<Project>
+  <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />
   <Import Project="$(ProjectDir)..\..\Build\Common.proj" />
   <PropertyGroup>
-    <SchemaVersion>2.0</SchemaVersion>
+    <TargetFrameworks>net45</TargetFrameworks>
+    <PlatformTarget>anycpu</PlatformTarget>
     <ProjectGuid>{4A617A40-2BA7-4713-AAFE-F90C4325C581}</ProjectGuid>
     <OutputType>Library</OutputType>
-    <OutputPath>$(OutputPath)\DLLs</OutputPath>
```
Comment (slozier) [inline Src/IronPython.SQLite/IronPython.SQLite.csproj]: If you remove this then IronPython.SQLite.dll is no longer in the DLLs subfolder.

File: Src/IronPython.Wpf/IronPython.Wpf.csproj [hunk 2]
```diff
@@ -1,40 +1,23 @@
...
+﻿<Project>
+  <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />
   <Import Project="$(ProjectDir)..\..\Build\Common.proj" />
   <PropertyGroup>
-    <SchemaVersion>2.0</SchemaVersion>
+    <TargetFrameworks>net45</TargetFrameworks>
+    <PlatformTarget>anycpu</PlatformTarget>
     <ProjectGuid>{65E997B7-E99B-4C83-B29E-9951429BB293}</ProjectGuid>
     <OutputType>Library</OutputType>
-    <OutputPath>$(OutputPath)\DLLs</OutputPath>
```
Comment (slozier) [inline Src/IronPython.Wpf/IronPython.Wpf.csproj]: Same here, this assembly used to be in the DLLs subfolder.

File: Src/IronPython/IronPython.csproj [hunk 3]
```diff
@@ -1,511 +1,85 @@
...
-      <Link>Properties\AssemblyVersion.cs</Link>
-    </Compile>
+    <ProjectReference Include="..\DLR\Src\Microsoft.Dynamic\Microsoft.Dynamic.csproj" />
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Include="..\AssemblyVersion.cs" Link="Properties\AssemblyVersion.cs" />
   </ItemGroup>
   <ItemGroup>
-    <Content Include="Lib\site.py">
+    <None Update="Lib\iptest\assert_util.py">
```
Comment (slozier) [inline Src/IronPython/IronPython.csproj]: Use a glob pattern instead of specifying each Lib file?

File: Src/IronPython/IronPython.csproj [hunk 4]
```diff
@@ -1,511 +1,85 @@
...
-    <EmbeddedResource Include="Resources.resx">
-      <SubType>Designer</SubType>
-      <Generator>ResXFileCodeGenerator</Generator>
-      <LastGenOutput>Resources.Designer.cs</LastGenOutput>
-    </EmbeddedResource>
+    <!-- if the file does not exist it's not picked up automatically -->
+    <Compile Remove="Properties\CurrentVersion.Generated.cs" />
+    <Compile Include="Properties\CurrentVersion.Generated.cs" />
+    <Compile Remove="Properties\BuildInfo.Generated.cs" />
+    <None Remove="Modules\unicodedata\IPyUnicodeData.txt.gz" />
```
Comment (slozier) [inline Src/IronPython/IronPython.csproj]: Do we need the remove here? If so maybe move it to next to the include of same file.

File: Src/IronPythonConsoleAny/IronPythonConsoleAny.csproj [hunk 5]
```diff
@@ -13,45 +14,26 @@
...
-    </ProjectReference>
-    <ProjectReference Include="..\IronPython\IronPython.csproj">
-      <Project>{95289EA9-5778-489D-AB48-F81F2CE2DA32}</Project>
-      <Name>IronPython</Name>
-    </ProjectReference>
+    <ProjectReference Include="..\IronPython\IronPython.csproj" />
   </ItemGroup>
-  <ItemGroup />
   <ItemGroup>
-    <Compile Include="..\IronPythonConsole\Properties\AssemblyInfo.cs">
```
Comment (slozier) [inline Src/IronPythonConsoleAny/IronPythonConsoleAny.csproj]: Where did this one go?

File: Src/IronPythonConsoleAny/IronPythonConsoleAny.csproj [hunk 6]
```diff
@@ -13,45 +14,26 @@
...
+    <Compile Include="..\AssemblyVersion.cs" Link="AssemblyVersion.cs" />
   </ItemGroup>
+  <Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk" />
+  <Import Project="$(BeforeTargetFiles)" />
+  <Target Name="BeforeBuild" DependsOnTargets="$(BeforeTargets);GenerateCurrentVersion" />
   <ItemGroup>
-    <Content Include="..\IronPythonConsole\ipy.ico" />
+    <!-- if the file does not exist it's not picked up automatically -->
+    <Compile Remove="Properties\CurrentVersion.Generated.cs" />
+    <Compile Include="..\IronPythonConsole\Console.cs" Link="Console.cs" />
```
Comment (slozier) [inline Src/IronPythonConsoleAny/IronPythonConsoleAny.csproj]: Maybe move this so it's not in the middle of the hack.

File: Src/IronPythonWindowAny/IronPythonWindowAny.csproj [hunk 7]
```diff
@@ -13,48 +14,29 @@
...
-      <Project>{02ff0909-f5ad-48cf-a86a-345e721b7e40}</Project>
-      <Name>Microsoft.Scripting</Name>
-    </ProjectReference>
-    <ProjectReference Include="..\IronPython\IronPython.csproj">
-      <Project>{95289EA9-5778-489D-AB48-F81F2CE2DA32}</Project>
-      <Name>IronPython</Name>
-    </ProjectReference>
+    <!-- if the file does not exist it's not picked up automatically -->
+    <Compile Remove="Properties\CurrentVersion.Generated.cs" />
+    <Compile Include="..\IronPythonWindow\IronPythonWindow.cs" Link="IronPythonWindow.cs" />
```
Comment (slozier) [inline Src/IronPythonWindowAny/IronPythonWindowAny.csproj]: Move to somewhere else in csproj file.

File: Src/IronPythonWindowAny/IronPythonWindowAny.csproj [hunk 8]
```diff
@@ -13,48 +14,29 @@
     <Prefer32Bit>false</Prefer32Bit>
   </PropertyGroup>
   <ItemGroup>
-    <Compile Include="..\IronPythonWindow\IronPythonWindow.cs" />
-    <Compile Include="..\IronPythonWindow\Properties\AssemblyInfo.cs">
```
Comment (slozier) [inline Src/IronPythonWindowAny/IronPythonWindowAny.csproj]: Where is this one?

File: packages.config [hunk 9]
```diff
@@ -1,5 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
```
Comment (slozier) [inline packages.config]: Can you add this back in for now? There's some associated cleanup that should be done with its removal.

File: Src/IronPython.SQLite/IronPython.SQLite.csproj (hunk 1 above)
Comment (gpetrou) [inline Src/IronPython.SQLite/IronPython.SQLite.csproj]: Sorry, why do we need to have a separate DLLs subfolder? Does it have to do with the library distribution?

Comment (slide) [inline Src/IronPython.SQLite/IronPython.SQLite.csproj]: Yes, IronPython looks in there for assemblies to load and look for modules. It's required.

**Comment by slozier** [comment]
Travis is not happy about something, maybe it's missing a "dotnet restore"?

File: make.cmd [hunk 10]
```diff
@@ -126,14 +126,14 @@ msbuild /t:DistClean /p:BuildFlavour=Debug /verbosity:minimal /nologo /p:Platfor
...
-%SystemRoot%\Microsoft.NET\Framework\v4.0.30319\ngen.exe install bin\Release\ipy.exe
+%SystemRoot%\Microsoft.NET\Framework\v4.0.30319\ngen.exe install bin\Release\net45\ipy.exe
 goto :exit
 
 :main
 msbuild Build.proj /t:%_target% /p:BuildFlavour=%_flavour% /verbosity:minimal /nologo /p:Platform="Any CPU"
 if "%_target%" == "Build" (
     echo "Copying %_flavour% test dlls"
-    xcopy /y /q Src\DLR\bin\%_flavour%\*.dll bin\%_flavour%\
+    xcopy /y /q Src\DLR\bin\%_flavour%\net45\*.dll bin\%_flavour%\
```
Comment (slozier) [inline make.cmd]: The destination probably needs to be `bin\%_flavour%\net45`. Not a great solution for multi-targetting but unless someone has a better idea to deploy the DLLs I suggest we fix it in a separate PR.

File: Makefile [hunk 11]
```diff
@@ -4,11 +4,11 @@ MSBUILD := msbuild
 
 release: update-submodules
 	@$(MSBUILD) Build.proj /t:Build /p:Mono=true /p:BuildFlavour=Release /p:Platform="Any CPU" /verbosity:minimal /nologo
-	cp Src/DLR/bin/Release/rowantest.*.dll bin/Release/
+	cp Src/DLR/bin/Release/net45/rowantest.*.dll bin/Release/
```
Comment (slide) [inline Makefile]: Don't these need to be copied to bin/Release/net45?

File: Src/IronPython/IronPython.csproj [hunk 12]
```diff
@@ -483,18 +28,26 @@
...
   <ItemGroup>
-    <ProjectReference Include="..\DLR\Src\Microsoft.Dynamic\Microsoft.Dynamic.csproj">
-      <Project>{eb66b766-6354-4208-a3d4-aacbdcb5c3b3}</Project>
-      <Name>Microsoft.Dynamic</Name>
-    </ProjectReference>
-    <ProjectReference Include="..\DLR\Src\Microsoft.Scripting\Microsoft.Scripting.csproj">
-      <Project>{02ff0909-f5ad-48cf-a86a-345e721b7e40}</Project>
-      <Name>Microsoft.Scripting</Name>
-    </ProjectReference>
+    <None Update="Lib\**\*.py">
```
Comment (slide) [inline Src/IronPython/IronPython.csproj]: Does this maintain the hierarchy during copy, such that Lib/iptest/*.py will be in OutputDirectory/Lib/iptest/*.py?

Comment (slozier) [inline Src/IronPython/IronPython.csproj]: Yes, the hierarchy will be maintained.

**Comment by gpetrou** [comment]
Any ideas what the stack overflow error is related to?

**Comment by slozier** [comment]
StackOverflow: not sure what the exact cause is, but we can disable the test for now...

**Comment by slide** [comment]
Change it to PROCESS Isolation in the .ini file, then it will just fail and not bring down the whole testing process.

**Pull request merged**